drivers: pinctrl: r8a779f0: move pfc file from soc to driver dir

xakep-amatop opened the pull request:
Move pfc_r8a779f0.c from the 'soc' directory to the 'drivers' directory to avoid code duplication for the ARM64 Spider board.

xakep-amatop committed:
drivers: pinctrl: r8a779f0: move pfc file from soc to driver dir

Move pfc_r8a779f0.c from the 'soc' directory to the 'drivers'
directory to avoid code duplication for the ARM64 Spider board.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

gmarull commented on drivers/pinctrl/CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ zephyr_library_sources_ifdef(CONFIG_PINCTRL_ITE_IT8XXX2 pinctrl_ite_it8xxx2.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_NPCX pinctrl_npcx.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_NUMICRO pinctrl_numicro.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_NRF pinctrl_nrf.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_R8A779F0 pfc_r8a779f0.c)
```
why `CONFIG_SOC_...` down here?

xakep-amatop replied:
Because all the code inside the file is related to the specified SoC, but the pfc rcar driver can be used for other RCar SoCs as well. Do you mean that it would be better to enclose all code inside the file with `#ifdef CONFIG_SOC_R8A779F0`?

As far as I can see, some other drivers use a similar approach, such as in the following examples:

```
bluetooth/hci/CMakeLists.txt:28:zephyr_library_sources_ifdef(CONFIG_SOC_NRF5340_CPUAPP nrf53_support.c)

flash/CMakeLists.txt:69:  if(CONFIG_SOC_SERIES_STM32H7X)
flash/CMakeLists.txt:71:  elseif(CONFIG_SOC_SERIES_STM32WBAX)

clock_control/CMakeLists.txt:37:if(CONFIG_SOC_SERIES_STM32MP1X)
clock_control/CMakeLists.txt:45:elseif(CONFIG_SOC_SERIES_STM32WBAX)
clock_control/CMakeLists.txt:49:  zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_STM32C0X clock_stm32c0.c)
clock_control/CMakeLists.txt:50:  zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_STM32F0X clock_stm32f0_f3.c)
clock_control/CMakeLists.txt:62:  zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_STM32WBX clock_stm32l4_l5_wb_wl.c)

clock_control/CMakeLists.txt:63:  zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_STM32WLX clock_stm32l4_
```

aaillet commented:
> Move pfc_r8a779f0.c from the 'soc' directory to the 'drivers' directory to avoid code duplication for the ARM64 Spider board.

I'm not sure that it's authorized to move such code in driver part, you should get general arm and pinctrl maintainers for their inputs.

On renesas part,
It sound nice to be able to use same code for both ARM32 and ARM64 part of the SoC if possible.
But, in this spirit, everything should move to driver folder in an all in one PR (one commit per soc) for it to be a complete work.
- Both `pfc_r8a77951.c` (duplicated) and `pfc_r8a77961.c` (arm64 part only is supported) are good candidate to move to driver folder.
- Common pinctrl header located in `zephyr/soc/arm/renesas_rcar/common/pinctrl_rcar.h` that is replacing both gen3 and gen4 `pinctrl_soc.h` should also move to driver part (or in it's `zephyr/include/zephyr/dt-bindings/pinctrl/` counterpart ?) and would be the only necessary custom header for rcar pfc drivers.

For all files to be moved, file name could change to meet what is done in the target folder of course.

xakep-amatop committed:
soc: renesas_rcar: make common pinctrl header for both arm and arm64 archs

Make common pinctrl header for both arm and arm64 archs in order to avoid
code duplicates.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented:
> > Move pfc_r8a779f0.c from the 'soc' directory to the 'drivers' directory to avoid code duplication for the ARM64 Spider board.
> 
> I'm not sure that it's authorized to move such code in driver part, you should get general arm and pinctrl maintainers for their inputs.
> 
> On renesas part, It sound nice to be able to use same code for both ARM32 and ARM64 part of the SoC if possible. But, in this spirit, everything should move to driver folder in an all in one PR (one commit per soc) for it to be a complete work.
> 
>     * Both `pfc_r8a77951.c` (duplicated) and `pfc_r8a77961.c` (arm64 part only is supported) are good candidate to move to driver folder.
> 
>     * Common pinctrl header located in `zephyr/soc/arm/renesas_rcar/common/pinctrl_rcar.h` that is replacing both gen3 and gen4 `pinctrl_soc.h` should also move to driver part (or in it's `zephyr/include/zephyr/dt-bindings/pinctrl/` counterpart ?) and would be the only necessary custom header for rcar pfc drivers.
> 
> 
> For all files to be moved, file name could change to meet what is done in the target folder of course.

done

gmarull commented:
> > > > Move pfc_r8a779f0.c from the 'soc' directory to the 'drivers' directory to avoid code duplication for the ARM64 Spider board.
> > > 
> > > 
> > > I'm not sure that it's authorized to move such code in driver part, you should get general arm and pinctrl maintainers for their inputs.
> > > On renesas part, It sound nice to be able to use same code for both ARM32 and ARM64 part of the SoC if possible. But, in this spirit, everything should move to driver folder in an all in one PR (one commit per soc) for it to be a complete work.
> > > ```
> > > * Both `pfc_r8a77951.c` (duplicated) and `pfc_r8a77961.c` (arm64 part only is supported) are good candidate to move to driver folder.
> > > 
> > > * Common pinctrl header located in `zephyr/soc/arm/renesas_rcar/common/pinctrl_rcar.h` that is replacing both gen3 and gen4 `pinctrl_soc.h` should also move to driver part (or in it's `zephyr/include/zephyr/dt-bindings/pinctrl/` counterpart ?) and would be the only necessary custom header for rcar pfc drivers.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > For all files to be moved, file name could change to meet what is done in the target folder of course.
> > 
> > 
> > done
> 
> Thanks for quick fix.
> 
> @gmarull as pinctrl maintainer, you may want us to put these files in a renesas folder in the pinctrl driver folder as it is done in Linux (see https://elixir.bootlin.com/linux/latest/source/drivers/pinctrl/renesas). Decision was taken not to create folder for zephyr clock drivers. It was not exactly the same case as device trees are directly referring to soc specific clock drivers but it may be a good indicator to take a decision.

I'm fine with a folder.

xakep-amatop commented:
> > > > Move pfc_r8a779f0.c from the 'soc' directory to the 'drivers' directory to avoid code duplication for the ARM64 Spider board.
> > > 
> > > 
> > > I'm not sure that it's authorized to move such code in driver part, you should get general arm and pinctrl maintainers for their inputs.
> > > On renesas part, It sound nice to be able to use same code for both ARM32 and ARM64 part of the SoC if possible. But, in this spirit, everything should move to driver folder in an all in one PR (one commit per soc) for it to be a complete work.
> > > ```
> > > * Both `pfc_r8a77951.c` (duplicated) and `pfc_r8a77961.c` (arm64 part only is supported) are good candidate to move to driver folder.
> > > 
> > > * Common pinctrl header located in `zephyr/soc/arm/renesas_rcar/common/pinctrl_rcar.h` that is replacing both gen3 and gen4 `pinctrl_soc.h` should also move to driver part (or in it's `zephyr/include/zephyr/dt-bindings/pinctrl/` counterpart ?) and would be the only necessary custom header for rcar pfc drivers.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > For all files to be moved, file name could change to meet what is done in the target folder of course.
> > 
> > 
> > done
> 
> Thanks for quick fix.
> 
> @gmarull as pinctrl maintainer, you may want us to put these files in a renesas folder in the pinctrl driver folder as it is done in Linux (see https://elixir.bootlin.com/linux/latest/source/drivers/pinctrl/renesas). Decision was taken not to create folder for zephyr clock drivers. It was not exactly the same case as device trees are directly referring to soc specific clock drivers but it may be a good indicator to take a decision.
> 
> @xakep-amatop
> 
>     * Please rename `include/zephyr/drivers/pinctrl/pinctrl_rcar.h` to `pinctrl_renesas_rcar_common.h` or `pinctrl_rcar_common.h` to meet file names in this folder.
> 

I'll do it

>     * I think that `drivers/pinctrl/pfc_r8a77951_r8a77961.c` filename is not that nice. I think we should keep a file per soc (to follow how it's done in linux).
> 

I believe it's preferable to keep it unchanged to prevent code duplication. After reviewing all tables in pfc_r8a77951_r8a77961.c, I can confirm that all described pins are compatible with both SoCs. 

>     * We need these files to be tracked by our MAINTAINER area
>       
>       * so i think you should add `rcar` in file name (eg. `drivers/pinctrl/pfc_rcar_r8a77951.c`) to be included in tracked files
>       * or directly `drivers/pinctrl/renesas` if decision is taken to create a separate folder.

sure, I'll create a new directory

aaillet commented:
> > ```
> > * I think that `drivers/pinctrl/pfc_r8a77951_r8a77961.c` filename is not that nice. I think we should keep a file per soc (to follow how it's done in linux).
> > ```
> 
> I believe it's preferable to keep it unchanged to prevent code duplication. After reviewing all tables in pfc_r8a77951_r8a77961.c, I can confirm that all described pins are compatible with both SoCs.

I agree with you and indeed, these two lists are most likely the same, but this is also the case under linux and they have always preferred to keep one file per soc (for a question of readability I guess).

Since most of Renesas' work on linux is done by internal Renesas developers, I try to follow what they've done as best I can (they may have access to internal information that helps make these kinds of decisions).
Staying as close as possible to Linux also helps to maintain a certain logic, given that the two worlds are, in most cases, supposed to cohabit.
It also makes it easier to compare the two sources and switch from one to the other.
In any case, for the time being, most of my code tree decisions have come from what has been done under Linux, so please do the same - it's best not to mix different ways of doing things.

aaillet commented on drivers/pinctrl/renesas/pfc_r8a779f0.c:
```diff
@@ -6,7 +6,7 @@
 
 #include <errno.h>
-#include "pinctrl_soc.h"
+#include <pinctrl_soc.h>
```
Why moving to <> here but not in other socs pfc files ?

aaillet commented on boards/arm64/rcar_salvator_xs_m3/salvator_xs_m3-pinctrl.dtsi:
```diff
@@ -4,7 +4,7 @@
  */
 
-#include <zephyr/dt-bindings/pinctrl/renesas/pinctrl-r8a77961.h>
+#include <zephyr/dt-bindings/pinctrl/renesas/pinctrl-r8a77951.h>
```
As said in previous comment, I know it's creating code duplication, but I prefer this board to keep only references to "r8a77961" soc.
Mixing references to both "r8a77951" & "r8a77961" may confused anyone reading board description.

So please, as in Linux, keep `include/zephyr/dt-bindings/pinctrl/renesas/pinctrl-r8a77961.h` and the reference to it here.

(ref. https://elixir.bootlin.com/linux/latest/source/drivers/pinctrl/renesas/pfc-r8a7796.c)

xakep-amatop replied:
done

aaillet commented on MAINTAINERS.yml:
What do you think of adding sub directories to the `drivers/pinctrl/renesas/` directory for each area to refer to ?
It would match effort started here (https://github.com/zephyrproject-rtos/zephyr/pull/67487/commits/b7bdfc934257e3ffb076dc06f0234be3fa794e1c) to organize dts folders.

xakep-amatop replied:
created subdirs

aaillet commented on soc/arm64/renesas_rcar/gen3/pinctrl_soc.h:
```diff
@@ -1,119 +1,11 @@
 /*
- * Copyright (c) 2021 IoT.bzh
+ * Copyright (c) 2024 EPAM Systems
```
please keep both copyright

```suggestion
 * Copyright (c) 2021 IoT.bzh
 * Copyright (c) 2024 EPAM Systems
```

xakep-amatop replied:
done

aaillet reviewed (changes requested):
LGTM overall, some more changes to request

aaillet commented on drivers/pinctrl/renesas/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+zephyr_library_sources_ifdef(CONFIG_PINCTRL_RZT2M rz/pinctrl_rzt2m.c)
+zephyr_library_sources_ifdef(CONFIG_PINCTRL_SMARTBOND smartbond/pinctrl_smartbond.c)
+
+add_subdirectory(rcar)
```
Can it be conditionally added depending on `CONFIG_PINCTRL_RCAR_PFC` ?

aaillet commented on drivers/pinctrl/renesas/rcar/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+zephyr_library_sources_ifdef(CONFIG_SOC_R8A77961 pfc_r8a77961.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_R8A779F0 pfc_r8a779f0.c)
+
+zephyr_library_sources_ifdef(CONFIG_PINCTRL_RCAR_PFC pfc_rcar.c)
```
No need for ifdef here ?
Move it as first include using `zephyr_library_sources` as it is the common one.

aaillet commented on drivers/pinctrl/renesas/rcar/pfc_r8a77961.c:
Shouldn't it be a file creation instead of a move of `soc/arm64/renesas_rcar/gen3/pfc_r8a77951.c` ?

xakep-amatop replied:
Nope, as I wrote before, currently, all tables are identical. There are definitely some different pins, but in `pfc_r8a77951.c`, they aren't referenced. So, we could use the same table as in `pfc_r8a77951.c`. See the answer here: https://github.com/zephyrproject-rtos/zephyr/pull/67357#issuecomment-1889259006

aaillet reviewed (changes requested):
Thanks for quick fixes, some more suggestions.
I'll test your patches on board ASAP too.

aaillet commented on MAINTAINERS.yml:
```diff
@@ -3395,6 +3395,7 @@ Renesas SmartBond Platforms:
     - dts/arm/renesas/smartbond/
     - dts/bindings/*/renesas,smartbond*
     - soc/arm/renesas_smartbond/
+    - drivers/pinctrl/renesas/smartbond/
```
Please add drivers related area under existing driver area

```suggestion
    - drivers/pinctrl/renesas/smartbond/
    - dts/arm/renesas/smartbond/
    - dts/bindings/*/renesas,smartbond*
    - soc/arm/renesas_smartbond/
```

xakep-amatop committed:
drivers: pinctrl: move renesas pincontrol drivers to separate directory

Moved r8a77951/r8a77961 pfc tables and funcs from soc to drivers/pfc dir.

Move all Renesas pinctrl driver source and header files to a separate
directory "renesas" and every SoC series to a separate subdirectory.

Update the MAINTAINERS file to align with the changes in the
pinctrl directory.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented:
@aaillet @gmarull pls, review changes again

aaillet commented on drivers/pinctrl/renesas/rcar/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2024 EPAM Systems
+# SPDX-License-Identifier: Apache-2.0
+
+if (CONFIG_SOC_R8A77951 OR CONFIG_SOC_ARM64_R8A77951)
+  zephyr_library_sources(pfc_r8a77951.c)
+endif()
+
+zephyr_library_sources_ifdef(CONFIG_SOC_R8A77961 pfc_r8a77961.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_R8A779F0 pfc_r8a779f0.c)
+
+zephyr_library_sources(pfc_rcar.c)
```
please move global source first

```suggestion
zephyr_library_sources(pfc_rcar.c)

if (CONFIG_SOC_R8A77951 OR CONFIG_SOC_ARM64_R8A77951)
  zephyr_library_sources(pfc_r8a77951.c)
endif()

zephyr_library_sources_ifdef(CONFIG_SOC_R8A77961 pfc_r8a77961.c)
zephyr_library_sources_ifdef(CONFIG_SOC_R8A779F0 pfc_r8a779f0.c)
```

xakep-amatop replied:
I see no reason for doing this I just save the same building list as it was before socs files first and then files from driver dir

aaillet replied:
It was a nit-picks, trying to sort things.
Global -> specific

Thanks

aaillet commented on MAINTAINERS.yml:
```diff
@@ -3422,6 +3425,7 @@ Renesas R-Car Platforms:
     - boards/arm/rcar_*/
     - boards/arm64/rcar_*/
     - drivers/*/*rcar*
+    - drivers/pinctrl/renesas/rcar/
     - drivers/clock_control/*cpg_mssr*
```
Please sort in alphabetical order (as in file tree)

```suggestion
    - drivers/clock_control/*cpg_mssr*
    - drivers/pinctrl/renesas/rcar/
```

xakep-amatop replied:
Could you please let me know when you finish the review? I am having difficulty understanding whether you have completed the review or just posted suggestions from your partial review.

I'm concerned that after fixing this one cosmetic issue, you may discover a new cosmetic issue in the unchanged code after a few days or weeks

aaillet replied:
Review of changes is finished, it was last few nit-picks, I forgot to stale my "request for changes"

I'm only reviewing changed code.

xakep-amatop commented:
fixed comments and rebased

aaillet reviewed (dismissed):
LGTM now, thanks for the fixes

xakep-amatop commented:
@gmarull Could you please review this PR again?

nordicjm reviewed (dismissed):
Needs converting to hwmv2

xakep-amatop commented:
> Needs converting to hwmv2

done
